tests/test_base_virtual_machine.cpp

```diff
@@ -104,36 +110,101 @@ namespace
+    decltype(MOCK(ssh_channel_request_exec)) request_exec{MOCK(ssh_channel_request_exec)};
+    decltype(MOCK(ssh_userauth_publickey)) userauth_publickey{MOCK(ssh_userauth_publickey)};
+    decltype(MOCK(ssh_channel_is_closed)) channel_is_closed{MOCK(ssh_channel_is_closed)};
     const mpt::DummyKeyProvider key_provider{"keeper of the seven keys"};
 };
+
+struct IpExecution : public BaseVM, public WithParamInterface<std::tuple<int, std::string, std::vector<std::string>>>
+{
+};
 } // namespace
 
 TEST_F(BaseVM, get_all_ipv4_works_when_ssh_throws_opening_a_session)
 {
-    mpt::StubBaseVirtualMachine base_vm;
+    mpt::StubBaseVirtualMachine base_vm(mp::VirtualMachine::State::running);
 
     REPLACE(ssh_new, []() { return nullptr; }); // This makes SSH throw when opening a new session.
-    EXPECT_THROW(mp::SSHSession("theanswertoeverything", 42), mp::SSHException); // Test that it indeed does.
 
-    auto ipv4_count = base_vm.get_all_ipv4(key_provider);
-    EXPECT_EQ(ipv4_count.size(), 0u);
+    auto ip_list = base_vm.get_all_ipv4(key_provider);
+    EXPECT_EQ(ip_list.size(), 0u);
 }
 
 TEST_F(BaseVM, get_all_ipv4_works_when_ssh_throws_executing)
 {
-    mpt::StubBaseVirtualMachine base_vm;
+    mpt::StubBaseVirtualMachine base_vm(mp::VirtualMachine::State::running);
 
     // Make SSH throw when trying to execute something.
-    REPLACE(ssh_connect, [](auto...) { return SSH_OK; });
-    REPLACE(ssh_is_connected, [](auto...) { return true; });
-    REPLACE(ssh_channel_open_session, [](auto...) { return SSH_OK; });
-    REPLACE(ssh_channel_request_exec, [](auto...) { return SSH_ERROR; });
-    REPLACE(ssh_userauth_publickey, [](auto...) { return SSH_OK; });
-
-    // Check that it indeed throws at execution.
-    mp::SSHSession session{"host", 42};
-    EXPECT_THROW(session.exec("dummy"), mp::SSHException);
-
-    auto ipv4_count = base_vm.get_all_ipv4(key_provider);
-    EXPECT_EQ(ipv4_count.size(), 0u);
+    request_exec.returnValue(SSH_ERROR);
+
+    auto ip_list = base_vm.get_all_ipv4(key_provider);
+    EXPECT_EQ(ip_list.size(), 0u);
 }
+
+TEST_F(BaseVM, get_all_ipv4_works_when_instance_is_off)
+{
+    mpt::StubBaseVirtualMachine base_vm(mp::VirtualMachine::State::off);
+
+    EXPECT_EQ(base_vm.get_all_ipv4(key_provider).size(), 0u);
+}
+
+TEST_P(IpExecution, get_all_ipv4_works_when_ssh_works)
+{
+    mpt::StubBaseVirtualMachine base_vm(mp::VirtualMachine::State::running);
+
+    auto [ip_exit_status, ip_output, expected_ips] = GetParam();
+    auto remaining = ip_output.size();
+
+    REPLACE(ssh_channel_get_exit_status, [](auto...) { return SSH_OK; });
+
+    ssh_channel_callbacks callbacks{nullptr};
+    auto add_channel_cbs = [&callbacks](ssh_channel, ssh_channel_callbacks cb) {
+        callbacks = cb;
+        return SSH_OK;
+    };
+    REPLACE(ssh_add_channel_callbacks, add_channel_cbs);
+
+    auto event_dopoll = [&callbacks, &ip_exit_status](ssh_event, int timeout) {
```

Comment:
This should work
```suggestion
    auto event_dopoll = [&callbacks, ip_exit_status = ip_exit_status](ssh_event, int timeout) {
```

It implies copying into new variables but maybe that's ok. Otherwise you'd need to get to declare the tuple explicitly and capture that:

```C++
auto params = GetParam();
auto mylambda = [&params] {... };
```

In that case I think a struct might be in order, to replace the tuple and have explicit names instead of `get<0>`.

Reply:
Amazing, thanks for the investigation! I implemented what you propose, using a `struct` is indeed much clearer. (It worked with clang, but for gcc I had to move the struct definition outside of the unnamed namespace.)